### PR TITLE
fix(react-router): resolve staticData in createFileRoute

### LINKED ISSUE #1978: Static route data is not enforced
### Describe the bug

When using `createFileRoute`, the option `staticData` is always optional.

### Your Example Website or App

https://stackblitz.com/edit/github-vnv6ze?file=src%2Froutes%2Findex.tsx

### Steps to Reproduce the Bug or Issue

1. Declare interface `StaticDataRouteOption` in `main.tsx`

### Expected behavior

The property `staticData` should be required when `StaticDataRouteOption` has any required properties.

### Screenshots or Videos

_No response_

### Platform

- OS: Windows
- Browser: Edge
- Version: 126.0.2592.102

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by honzahruby**
Resolves #1978 
`createFileRoute` didn't correctly resolve and enforce `staticData` even when it should. The return type for this function was always compiled down with typescript instead of using the "live" types. During the library build, the `StaticDataRouteOption` was empty, and it resulted in the `staticData` being always optional.

Screenshot 2024-08-02 at 15 04 59
Screenshot 2024-08-02 at 15 05 29
Screenshot 2024-08-02 at 15 06 29

### CODE REVIEW

**Comment by honzahruby** [comment]
Looks like this somehow breaks inferring the router context in `beforeLoad` 😕

**Comment by SeanCassiere** [comment]
> #1978

Could you clarify? Testing this locally, I can't see the beforeLoad inference breaking. Are you using `createRootRoute` or `createRootRouteWithContext`?

**Review by SeanCassiere — CHANGES_REQUESTED** [review]
Putting a RC block on this for now till my comment gets resolved.

Testing locally, I couldn't really observe any breakages and it does indeed fix the linked issue.

Once resolved, we can approve and merge this in.

**Comment by honzahruby** [comment]
@SeanCassiere My bad, I mixed two different versions of the router. This change seems to work fine in the `kitchen-sink-file-based` example, let's move on with it 🚀

**Review by SeanCassiere — APPROVED** [review]
Will check it again and merge in an hour.

**Comment by schiller-manuel** [comment]
a type test would have been nice, can you add one separately?

**Comment by honzahruby** [comment]
@schiller-manuel Absolutely, I wanted to add some but there weren’t any even for the basic createRoute. I’ll have a look at it 👍

**Comment by honzahruby** [comment]
Since `StaticDataRouteOption` is a global interface, it's available to all test files. If we add any required prop to staticData, we'd have to change all existing tests. Any idea how to test this in isolation? 🤔

**Pull request merged**
